rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,377 @@
+| File size             | the size of the file to be uploaded                                        |
+| File Name             | the name of the file being uploaded                                        |
+| submission_id         | identifies the submission                                                  |
+| attestation           | A boolean value indicating if the file is for attestation                  |
+| _Optional_ dataset_id | identifies the dataset being uploaded. Used if resuming an existing upload |
+
+**Response:**
+
+| Key           | Description                                                   |
+| ------------- | ------------------------------------------------------------- |
+| Access Token  | allows the user to upload their file to S3 in multiple parts. |
+| Presigned URL | allows the user to upload their file to S3 in a single part.  |
+| dataset_id    | identifies the name of the new file created.                  |
+
+**Error Responses:**
+
+| Code | Description                                                                                                             |
+| ---- | ----------------------------------------------------------------------------------------------------------------------- |
+| 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |
+| 400  | if the files type is invalid                                                                                            |
+
+### Upload Flows
+
+There are two methods for uploading data to the DP. They are Cloud to Cloud and Local to Cloud. The follow
+sections details the user interactions required, and how the Data Portal backend will facilitate the upload.
+
+#### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their CSP from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+##### 1. Get Share Link
+
+The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or submission of files.
+
+##### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
```

Comment:
Do you have some examples you could share? What types of transforms will you perform on these links? How will these be sanitized?

Reply:
I'm open to any tools if you have any to suggest.

Currently, the URLs generated by Dropbox and Google Drive aren't direct links to the files you want to download. For dropbox you need change one of query parameters to tell it to download. For google drive you need to parse out the file id, and construct a new URL with that ID to download the file. We can match the domain, and use a regex to verify the URL is in the format we expect before trying to use it.